### PR TITLE
Feature/score evaluation optimizing

### PR DESCRIPTION
**Pull request by Dronplane**
Optimized conjunction and disjunction performance

### CODE REVIEW

File: core/search/conjunction.hpp [hunk 1]
```diff
@@ -125,25 +125,56 @@ class conjunction : public doc_iterator_base, score_ctx {
...
       const auto* score = it.score;
       if (&irs::score::no_score() != score) {
         scores_.push_back(score);
+        scores_vals_.push_back(score->c_str());
       }
     }
-
+   
     if (scores_.empty()) {
       prepare_score(ord, nullptr, [](const score_ctx*, byte_type*) { /*NOOP*/});
```
Comment (gnusi) [inline core/search/conjunction.hpp]: let's move it under `switch`

File: core/search/disjunction.hpp [hunk 2]
```diff
@@ -80,6 +81,18 @@ void score_add(byte_type* dst, const order::prepared& order, DocIterator& src) {
   add_score_fns[&irs::score::no_score() == score](order, *score, dst);
 }
 
+template<typename DocIterator>
+void evaluate_score_iter(const irs::byte_type**& pVal,  DocIterator& src) {
+  const auto* score = src.score;
+  assert(score);
+  if (&irs::score::no_score() != score) {
+    score->evaluate();
+    *pVal = score->c_str();
```
Comment (gnusi) [inline core/search/disjunction.hpp]:
```suggestion
*pVal++ = score->c_str();
```

File: core/search/disjunction.hpp [hunk 3]
```diff
@@ -351,7 +371,7 @@ class small_disjunction : public doc_iterator_base, score_ctx {
         scored_itrs_.emplace_back(it);
       }
     }
-
+    scores_vals_.resize(scored_itrs_.size(), nullptr);
```
Comment (gnusi) [inline core/search/disjunction.hpp]:
```
scores_vals_.resize(scored_itrs_.size());
```

File: core/search/disjunction.hpp [hunk 4]
```diff
@@ -372,9 +391,11 @@ class small_disjunction : public doc_iterator_base, score_ctx {
 
           if (doc == self.doc_.value) {
             it.score->evaluate();
-            self.ord_->add(score, it.score->c_str());
+            *pVal = it.score->c_str();
```
Comment (gnusi) [inline core/search/disjunction.hpp]:
```suggestion
*pVal++ = it.score->c_str();
```

File: core/search/disjunction.hpp [hunk 5]
```diff
@@ -584,27 +606,28 @@ class disjunction : public doc_iterator_base, score_ctx {
...
       });
     }
+    ord_->merge(lhs, scores_vals_.data(), std::distance(scores_vals_.data(), pVal));
   }
 
   doc_iterators_t itrs_;
   std::vector<size_t> heap_;
   document doc_;
   const order::prepared* ord_;
+  mutable std::vector<const irs::byte_type*> scores_vals_;
```
Comment (gnusi) [inline core/search/disjunction.hpp]: move `std::vector<const irs::byte_type*>` closer to `std::vector<size_t>`

**Comment by Dronplane** [comment]
https://jenkins.arangodb.biz/view/PR/job/iresearch_s-pr/37/

**Pull request merged**